src/proc.cpp

```diff
@@ -362,27 +362,60 @@ static bool process_mark_finished_children(bool block_on_fg) {
     // This is always called from the main thread (and not forked), so we can cache this value.
     static pid_t shell_pgid = getpgrp();
 
+    // We can't always use SIGCHLD to determine if waitpid() should be called since it is not
+    // strictly one-SIGCHLD-per-one-child-exited (i.e. multiple exits can share a SIGCHLD call) and
+    // we a) return immediately the first time a dead child is reaped, b) explicitly skip over jobs
+    // that aren't yet fully constructed, so it's possible that we can get SIGCHLD and even find a
+    // killed child in the jobs we are reaping, but also have an exited child process in a job that
+    // hasn't been fully constructed yet - which means we can end up never knowing about the exited
+    // child process in that job if we use SIGCHLD count as the only metric for whether or not
+    // waitpid() is called.
+    // Without this optimization, the slowdown caused by calling waitpid() even just once each time
+    // `process_mark_finished_children()` is called is rather obvious (see the performance-related
+    // discussion in #5219), making it worth the complexity of this heuristic.
+
+    /// Tracks whether or not we received SIGCHLD without checking all jobs (due to jobs under
+    /// construction), forcing a full waitpid loop.
+    static bool dirty_state = true;
+    static size_t last_jobs_hash = -1;
```

Comment:
Is this used anywhere?

Reply:
Nope. As mentioned, that entire commit was part of a different optimization strategy and has now been dropped via rebase.